association_backend/association/domain/repositories/association_repository.py

```
@@ -0,0 +1,149 @@
+        # the actual Session ID is returned in the query parameter when your customer
+        # is redirected to the success page.
+        customer_payload = {}
+        if data.customer_id:
+            customer_payload.update(dict(customer=data.customer_id))
+        elif data.customer_email:
+            customer_payload.update(dict(customer_email=data.customer_email))
+        checkout_session_stripe_key = "{CHECKOUT_SESSION_ID}"
+        checkout_session = stripe.checkout.Session.create(
+            success_url=f"{STRIPE_SUBSCRIPTION_SUCCESS_URL}?session_id={checkout_session_stripe_key}",
+            cancel_url=STRIPE_SUBSCRIPTION_CANCEL_URL,
+            payment_method_types=["card"],
+            mode="subscription",
+            line_items=[
+                {
+                    "price": STRIPE_SUBSCRIPTION_PRICE_ID,
+                    # For metered billing, do not pass quantity
+                    "quantity": 1,
+                }
+            ],
+            **customer_payload,
+            api_key=STRIPE_SUBSCRIPTION_API_SECRET,
+        )
+        logger.info(f"checkout_session: {checkout_session}")
+        return StripeCheckoutSession(
+            id=checkout_session["id"],
+            customer_id=checkout_session["customer"] or "",
+            subscription_id=checkout_session["subscription"] or "",
+        )
+
+    async def retrieve_customer_portal_session_url(self, customer_id: str) -> str:
+        session = stripe.billing_portal.Session.create(
+            customer=customer_id,
+            return_url=DOMAIN_URL,
+            api_key=STRIPE_SUBSCRIPTION_API_SECRET,
+        )
+        return session.url
+
+    # READ FROM STRIPE
+    async def retrieve_customer_by_email(self, email: str) -> Optional[StripeCustomer]:
+        customers = stripe.Customer.list(
+            email=email, api_key=STRIPE_SUBSCRIPTION_API_SECRET
+        )
+        if len(customers):
+            return StripeCustomer(id=customers.data[0].id, email=email)
```

should we throw an error if we find more than one? instead of just taking the first one

---

I think not.
Do yoo want to log it on Sentry?

---

I think we should throw an error. 

It's a case that should never happen, and if it happens that a user has two customer profiles and we should investigate. Otherwise, we might end up choosing the wrong customer and subscribe the wrong person :)

---

ok
you can see the implementation in this commit : [0617246c](https://github.com/madEng84/pycon/commit/0617246cf627d476a40b3a4ccbb62fbdbbcecf30)